wip: multicore

Dirbaio opened the pull request:
You can select a core and debug it:

```
[dirbaio@mars probe-rs]$ target/debug/probe-rs-cli debug --chip nrf53 --core 0
>> halt
Core stopped at address 0x000005d8
....

[dirbaio@mars probe-rs]$ target/debug/probe-rs-cli debug --chip nrf53 --core 1
>> halt
Core stopped at address 0x010005c0
[...]
```

Flashing automatically selects the correct flash algo, RAM and core for the address range:

```
probe-rs-cli --  download --chip nrf53 ../blah/nrf53-app.elf
(...)
 INFO  probe_rs::flashing::loader   > chosen algo: RawFlashAlgorithm { name: "nrf53xx_application", description: "nRF53xxx_app", default: true, instructions: [b0, b5, 40, f2, 4, 5, xxxxx, 0, 0, 0, 0, 0, 0, 0, 0], pc_init: Some(1), pc_uninit: Some(41), pc_program_page: 135, pc_erase_sector: b5, pc_erase_all: Some(99), data_section_offset: 524, flash_properties: FlashProperties { address_range: 0..200000, page_size: 1000, erased_byte_value: ff, program_page_timeout: 3e8, erase_sector_timeout: bb8, sectors: [SectorDescription { size: 1000, address: 0 }] }, cores: ["application"] }
 INFO  probe_rs::flashing::loader   > chosen core: index 0, Core { name: "application", core_type: M33, arm_swd_ap: 0 }
 INFO  probe_rs::flashing::loader   > chosen RAM to run the algo: RamRegion { range: 20000000..20040000, is_boot_memory: false, cores: ["application"] }
(...)

probe-rs-cli --  download --chip nrf53 ../blah/nrf53-net.elf
(...)
 INFO  probe_rs::flashing::loader   > chosen algo: RawFlashAlgorithm { name: "nrf53xx_network", description: "nRF53xxx_net", default: true, instructions: [b0, b5, 40, f2,xxxxx, 0, 0, 0, 0, 0, 0, 0, 0, 0], pc_init: Some(1), pc_uninit: Some(41), pc_program_page: 135, pc_erase_sector: b5, pc_erase_all: Some(99), data_section_offset: 530, flash_properties: FlashProperties { address_range: 1000000..1040000, page_size: 800, erased_byte_value: ff, program_page_timeout: 3e8, erase_sector_timeout: bb8, sectors: [SectorDescription { size: 800, address: 0 }] }, cores: ["network"] }
 INFO  probe_rs::flashing::loader   > chosen core: index 1, Core { name: "network", core_type: M33, arm_swd_ap: 1 }
 INFO  probe_rs::flashing::loader   > chosen RAM to run the algo: RamRegion { range: 21000000..21020000, is_boot_memory: false, cores: ["network"] }
(...)
```


TODO list:
- [x] Use right core for RAM "pseudoflashing"
- [x] Update all YAMLs

noppej commented:
@Dirbaio ... I saw your comments on matrix about this PR, and want to offer my help. I have an H7 board, and will gladly make time available to help you test. It would also help me to collaborate with you to expand the `debugger` to handle multi-core. 
(cc: @Tiwalun , @Yatekii )

Yatekii commented on probe-rs/src/config/target.rs:
```diff
@@ -38,22 +38,23 @@ impl Target {
     /// Get the architectre of the target
     pub fn architecture(&self) -> Architecture {
-        match &self.core_type {
+        // TODO: what if they're different?
```
Could it even?

Dirbaio replied:
could be if the target yaml is wrong. It should probably be caught earlier, maybe in the "target validation" we once talked about

diondokter replied:
There are designs with different cores: https://www.st.com/en/microcontrollers-microprocessors/stm32wb55cg.html
This is an M4 & M0+

Yatekii commented on probe-rs-target/src/chip.rs:
```diff
@@ -28,3 +32,18 @@ pub struct Chip {
+    pub core_type: CoreType,
+
+    /// The AP number to access the core
+    pub arm_swd_ap: u8,
```
We should get rid of the `swd` portion here as the same API is available through JTAG too if JTAG is available.

Dirbaio replied:
Done

Dirbaio commented:
Rebased on latest master. NOTE: multicore flashing was previously working, with the rebase it's now WIP. Need to check how to adapt the new FlashBuilder code.

Added an example on how to do a recover on both cores (unlock): `cargo run --example nrf53_recover`

You can check if unlocked with `cargo run --bin probe-rs-cli -- info`. You should see two `Class1RomTable()` (the MEM-APs), and two `Unknown Type of access port` (these are the CTRL_APs)

diondokter commented:
Updated to-do:

- [ ] When loading target yamls, validate that all core names exist
- [x] Use right core for RAM "pseudoflashing"
- [ ] What if not all cores have the same architecture? YAML validation should probably fail..?
- [ ] Session still does debug_core_start on only core 0
- [ ] Update all YAMLs
- [ ] Update target-gen?
- [ ] Update the debugging code to respect core_index

Yatekii commented:
Different architectures are no issue yet as they only differentiate between ARM and RISC-V atm and those are always the same of either for each core on a chip.

diondokter commented:
All the target YAMLs have been added back. But any secondary cores still have to be added in

Dirbaio commented on probe-rs/targets/nRF53_Series.yaml:
```diff
@@ -5,10 +5,16 @@ variants:
+        core_access_options:
+          Arm:
+            ap: 1
+            psel: 0
```
What's PSEL?

diondokter replied:
Port select. @Yateki wanted that in too. It's for multi drop I think

Dirbaio replied:
Ah, nice! it should be a u32 then.

diondokter replied:
Ah! Changed it :)

Yatekii commented on gdb-server/src/architecture.rs:
```diff
@@ -141,7 +141,8 @@ impl GdbTargetExt for probe_rs::Target {
         // - riscv:rv32   -> RISCV
 
-        let architecture = match self.core_type {
+        // TODO: what if they're not all equal?
```
We should remove this todo and replace it with a comment. I'd assume they are always the same until we know better. If they ever differ I'd suggest that we extend this to return an array containing all the architectures.

I think I know of no MCU where this is an issue atm.

diondokter replied:
Ah, didn't spot this one. In one of my last commits I replaced it where something does pretty much the same thing with an assert. That way, at least we won't just assume that all cores have the same architecture and actually check it

Yatekii replied:
Sounds good!

Dirbaio replied:
The assert checks all have the same arch (arm vs riscv), but this piece of code cares about the actual version (armv6 vs armv7). There are chips out there with this. For example STM32WL is M4 + M0+. So the TODO still applies.

What do we do? I don't think GDB even supports debugging mixed-version chips like that. We can expose cores as "threads" to GDB, but GDB assumes they're all the same arch version.

Yatekii replied:
Ok, leave it as it is for now and revisit later. TBH I am all in favor of letting GDB die and go all in on our own DAP & debugging implementation once that hits beta/all basic features and then we do not have to care anymore.

Yatekii reviewed (changes requested):
Code looks good to me. I added a small nit and as discussed on Matrix, I will test this in a sec.

Yatekii commented:
Ok so all tests pass =)

The log of the run: https://gist.github.com/Yatekii/09eec906f586a760e2dbb751c3b54e05.

Before I merge I'd like to see the TODO fixed :)

Yatekii reviewed (approved):
We will merge this as it is now. The small issue on the GDB stup implementation will be left as a TODO as GDB most likely can't handle this anyways. So until it implements proper multicore (which probably will never happen as we have MDAP now) we can leave it as is.

I tested this PR against the nRF5340 as well and also fixed proper core<->memory mapping. To test stepping on core 1 one needs to write a register tho, which is not done yet. For this we need sequences, same as for unlock which was performed manually. Hence flashing was tested manually too.

Great work forks, thank you very much!

bors r+
